### PR TITLE
Fixed legend with legend location "best" when legend overlaps shaded area and text

### PR DESCRIPTION
**Pull request by alanlau28**
## PR summary

This PR fixes bug #27414 and #23323.  The calculation of badness in the legend location for 'best' now considers PolyCollection and Text. For PolyCollection, the paths are now converted to lines to check for hit detection when placing a legend. For Text, the bounding box is now checked for overlap.
## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: lib/matplotlib/legend.py [hunk 1]
```diff
@@ -983,6 +983,12 @@ def _auto_legend_data(self):
                 if len(hoffsets):
                     for offset in transOffset.transform(hoffsets):
                         offsets.append(offset)
+                if isinstance(artist, PolyCollection):
+                    paths = artist.get_paths()
+                    for path in paths:
+                        lines.append(artist.get_transform().transform_path(path))
```
Comment (ksunden) [inline lib/matplotlib/legend.py:989]: I would probably just have this in the if/elif chain rather than inside of one of the cases.

**Comment by alextanned** [comment]
> I do believe this should have a behavior change note, while "best" is intentionally 'we take care of it for you' and so I don't necessarily think it can't change, noting when the change occurs is something we should do.

I agree. May I ask how should I proceed with this? Should I add a sentence in the doc for loc='best'?

**Comment by alextanned** [comment]
> I do believe this should have a behavior change note, while "best" is intentionally 'we take care of it for you' and so I don't necessarily think it can't change, noting when the change occurs is something we should do.

I agree. How does the newly added note look?

**Comment by QuLogic** [comment]
Please see the [instructions in our contribution guide](https://matplotlib.org/devdocs/devel/contribute.html#new-changed-api).

**Comment by alextanned** [comment]
Hello, could you please provide some guidance on the failed tests? How should I address the [PR cleanliness](https://github.com/matplotlib/matplotlib/actions/runs/7152825818/job/19478584960?pr=27469#step:4:13) and [pytest failure](https://dev.azure.com/matplotlib/matplotlib/_build/results?buildId=37598&view=logs&j=aa39683c-d595-5f65-0b25-75adb2a2e3f3&t=2551f7a6-26d7-58a2-495a-78f1a540557c&l=214)? Thank you very much!

**Comment by rcomer** [comment]
The PR cleanliness failure is because you replaced a test image twice.  We don't want any more images in the commit history than necessary, so you will need to squash the commits down.

To squash, you can do an [interactive rebase](https://matplotlib.org/devdocs/devel/development_workflow.html#rewrite-commit-history).  I strongly recommend saving a backup copy of your branch first if you have not done this (much) before.

File: lib/matplotlib/legend.py [hunk 2]
```diff
@@ -978,11 +978,17 @@ def _auto_legend_data(self):
             elif isinstance(artist, Patch):
                 lines.append(
                     artist.get_transform().transform_path(artist.get_path()))
+            elif isinstance(artist, PolyCollection):
+                paths = artist.get_paths()
+                for path in paths:
+                    lines.append(artist.get_transform().transform_path(path))
```
[the remark this answers is not in the document]
Comment (oscargus) [reply lib/matplotlib/legend.py:984]: One may even consider replacing `paths` with `artist.get_paths()` for yet another minor optimization.

[the remark this answers is not in the document]
Comment (alanlau28) [reply lib/matplotlib/legend.py]: Tested the optimizations...Indeed a minor speedup! https://github.com/matplotlib/matplotlib/pull/27469/files#diff-e7a00bb4a353fa623c1d0c1de5c6ca602b7ab38e8906d4842640983133b9196aR981-R983

**Review by oscargus — APPROVED** [review]
Looks good to me! 

@alanlau28 if you can reduce the number of commits that would be great (rebase and fixup/squash)!  If not, we just have to remember to squash while merging.

**Comment by QuLogic** [comment]
Yes, we definitely do not want these "Empty-Commit" commits with nothing in them.

**Review by dstansby — APPROVED** [review]

**Comment by dstansby** [comment]
This looks good to go - it just needs rebasing on to the current `main` branch. @alanlau28 would you be able to do that? Let us know if you want any help, or someone else to do it.

**Comment by rcomer** [comment]
I think it would be good to get this in, so I took the liberty of rebasing and squashing.  The conflict was just because an adjacent branch of the if-loop was modified at #27517.

**Pull request merged**